knimin/lib/data_access.py

```diff
@@ -3532,6 +3533,204 @@ def read_dna_plate(self, dna_plate_id):
+                     VALUES (%s, now(), %s, %s)
+                     RETURNING shotgun_normalized_plate_id"""
+
+            TRN.add(sql, [shotgun_plate_id, email, echo_id])
+            shotgun_normalized_plate_id = TRN.execute_fetchindex()[0][0]
+
+            # add all of the volume details for the normalized plate
+            sql = """INSERT INTO pm.shotgun_normalized_plate_well_values
+                        (shotgun_normalized_plate_id, row, col,
+                         sample_volume_nl, water_volume_nl)
+                     VALUES (%s, %s, %s, %s, %s)"""
+            sql_args = []
+            for row in np.arange(n_rows):
+                for col in np.arange(n_cols):
+                    sql_args.append((shotgun_normalized_plate_id,
+                                     row, col,
+                                     plate_normalization_sample[row, col],
+                                     plate_normalization_water[row, col]))
+            TRN.add(sql, sql_args, many=True)
+            TRN.execute()
+        return shotgun_normalized_plate_id
+
+    def read_normalized_shotgun_plate(self, shotgun_normalized_plate_id):
+        """Obtain the normalized shotgun plate details
+
+        Parameters
+        ----------
+        shotgun_normalized_plate_id : int
+            The ID of the normalized shotgun plate to retrieve.
+
+        Raises
+        ------
+        ValueError
+            If the normalized_shotgun_plate_id does not exist
+
+        Returns
+        -------
+        dict
+            A dict populated by the columns details in
+            pm.shotgun_normalized_plate, as well as two numpy matrices
+            corresponding to the normalized water and sample volumes
+        """
+        with TRN:
+            # get plate details
+            sql = """SELECT *
+                     FROM pm.shotgun_normalized_plate
+                     WHERE shotgun_normalized_plate_id = %s"""
+            TRN.add(sql, [shotgun_normalized_plate_id])
+            res = TRN.execute_fetchindex()
+            if not res:
+                raise ValueError("normalized shotgun plate %s does not exist" %
+                                 shotgun_normalized_plate_id)
+
+            res = dict(res[0])
+            res.pop('shotgun_normalized_plate_id')  # it's redundant
+
+            # get plate shape
+            sql = """SELECT rows, cols
+                     FROM pm.shotgun_plate JOIN
+                          pm.plate_type USING(plate_type_id)
+                     WHERE shotgun_plate_id = %s"""
+            TRN.add(sql, [res['shotgun_plate_id']])
+            plate_shape = dict(TRN.execute_fetchindex()[0])
+            n_rows = plate_shape['rows']
+            n_cols = plate_shape['cols']
+            sample_volumes = np.zeros((n_rows, n_cols))
+            water_volumes = np.zeros((n_rows, n_cols))
+
+            # get well values
+            sql = """SELECT row, col, sample_volume_nl, water_volume_nl
```

Comment:
not seeing qpcr concentration in the schema, or qpcr cp. those details also seem to be part of `pm.shotgun_normalized_plate`? will add i5/i7